clang/lib/Parse/ParseStmt.cpp

```diff
@@ -2427,3 +2447,75 @@ bool Parser::ParseOpenCLUnrollHintAttribute(ParsedAttributes &Attrs) {
   }
   return true;
 }
+
+WhereClauseFact *Parser::ParseWhereClauseFact() {
+  SourceLocation Loc = Tok.getLocation();
+
+  // TODO: Handle bounds expression surrounded by parentheses, like:
+  // _Where ((((p : bounds(p, p + 1))))).
+  // Equality expressions surrounded by parentheses are already handled by
+  // invoking IgnoreValuePreservingCasts in ActOnEqualityOpFact.
+
+  if (Tok.is(tok::identifier) && NextToken().is(tok::colon)) {
+    IdentifierInfo *VarName = Tok.getIdentifierInfo();
+
+    // Consume the identifier.
+    ConsumeToken();
+    // Consume the ':' token.
+    ConsumeToken();
+
+    // Parse bounds expression.
+    ExprResult BoundsRes(ParseBoundsExpression());
+    if (BoundsRes.isInvalid())
+      return nullptr;
+    return Actions.ActOnBoundsFact(VarName, BoundsRes.get(),
+                                   getCurScope(), Loc);
+  }
+
+  ExprResult ExprRes = Actions.CorrectDelayedTyposInExpr(ParseExpression());
+  if (ExprRes.isInvalid())
+    return nullptr;
+  return Actions.ActOnEqualityOpFact(ExprRes.get(), Loc);
+}
+
+WhereClause *Parser::ParseWhereClause() {
+  // Consume the "_Where" token.
+  if (ExpectAndConsume(tok::kw__Where)) {
```

Comment:
I am using `ExpectAndConsume` here which will issue an error in case a `_Where` token is not seen.

I have removed `EmitDiag` since the parser and sema functions now handle errors themselves.